api/handlers.go

```diff
@@ -882,8 +882,381 @@ func (a *App) handleUnsubscribe(c *gin.Context) {
+	if err != nil {
+		errs.NewBadRequestError(pkgerr.Wrap(err, "id expects int64")).Abort(c)
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
+	db := c.MustGet("MY_DB").(*sql.DB)
+	err = sqlutil.InTx(context.TODO(), db, func(tx *sql.Tx) error {
+		b, err := models.FindBookmark(tx, id)
+
+		if err == sql.ErrNoRows {
+			return nil
+		}
+
+		if err != nil {
+			return pkgerr.Wrap(err, "fetch bookmark from db")
+		}
+		_, err = b.Delete(tx)
+		return err
+	})
+
+	concludeRequest(c, nil, err)
+}
+
+//Folder handlers
+func (a *App) handleGetFolders(c *gin.Context) {
+	var r GetFoldersRequest
+	if c.Bind(&r) != nil {
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
+	db := c.MustGet("MY_DB").(*sql.DB)
+
+	mods := []qm.QueryMod{models.FolderWhere.UserID.EQ(user.ID)}
+
+	total, err := models.Folders(mods...).Count(db)
+
+	_, offset := appendListMods(&mods, r.ListRequest)
+	if int64(offset) >= total {
+		concludeRequest(c, new(GetFoldersResponse), nil)
+		return
+	}
+
+	if r.BookmarkIdFilter != 0 {
+		mods = append(mods, qm.Load("BookmarkFolders", models.BookmarkFolderWhere.BookmarkID.EQ(r.BookmarkIdFilter)))
+	}
+	appendQueryFilter(&mods, r.QueryFilter, "name")
+	folders, err := models.Folders(mods...).All(db)
+	if err != nil {
+		errs.NewInternalError(pkgerr.WithStack(err)).Abort(c)
+		return
+	}
+
+	items := make([]*Folder, len(folders))
+	for i, f := range folders {
+		items[i] = makeFoldersDTO(f)
+	}
+
+	resp := GetFoldersResponse{
+		ListResponse: ListResponse{Total: total},
+		Items:        items,
+	}
+	concludeRequest(c, resp, err)
+}
+
+func (a *App) handleCreateFolder(c *gin.Context) {
+	var r AddFolderRequest
+	if c.BindJSON(&r) != nil {
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
+	db := c.MustGet("MY_DB").(*sql.DB)
+	folder := &models.Folder{UserID: user.ID}
+	if r.Name != "" {
+		folder.Name = null.StringFrom(r.Name)
+	}
+
+	err := sqlutil.InTx(context.TODO(), db, func(tx *sql.Tx) error {
+		err := folder.Insert(tx, boil.Infer())
+		if err != nil {
+			return pkgerr.WithStack(err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		errs.NewInternalError(pkgerr.WithStack(err)).Abort(c)
+		return
+	}
+	resp := makeFoldersDTO(folder)
+	concludeRequest(c, resp, err)
+}
+
+func (a *App) handleUpdateFolder(c *gin.Context) {
+	id, err := strconv.ParseInt(c.Param("id"), 10, 0)
+	if err != nil {
+		errs.NewBadRequestError(pkgerr.Wrap(err, "id expects int64")).Abort(c)
+		return
+	}
+
+	var r UpdateFolderRequest
+	if c.BindJSON(&r) != nil {
+		return
+	}
+
+	user := &models.User{}
+	if err := a.validateUser(c, user); err != nil {
+		err.Abort(c)
+		return
+	}
+
+	var resp *Folder
+	db := c.MustGet("MY_DB").(*sql.DB)
+	err = sqlutil.InTx(context.TODO(), db, func(tx *sql.Tx) error {
+		f, err := models.FindFolder(tx, id)
```

Comment:
again. need to verify this folder is actually owned by this user or this is a security breach

Reply:
fixed